prybar_assets/nodejs/input-sync.js

```diff
@@ -0,0 +1,158 @@
+const { readSync, writeSync, openSync } = require("fs");
+const { isatty } = require("tty");
+const buf = Buffer.alloc(1);
+const isTTY = isatty(process.stdin.fd);
+
+/**
+ * Reads a single byte from stdin to buf.
+ * @return {boolean} Whether a byte was read or not (false on EOF)
+ */
+function readByteSync() {
+  const read = readSync(stdinFd, buf);
+
+  return read > 0;
+}
+
+/**
+ * The file descriptor of our stdin stream.
+ * @type {number}
+ */
+const stdinFd = isTTY
+  ? // We can't just use process.stdin.fd here since node has some getter shenanigans
+    // which cause sync reads to throw
+    openSync("/dev/tty", "r")
+  : openSync("/dev/stdin", "r");
+
+/**
+ * Writes output to stdout.
+ *
+ * @param {NodeJS.ArrayBufferView} b The outpus which should be written to stdout.
+ */
+function writeOutput(b) {
+  writeSync(process.stdout.fd, b);
+}
+
+/**
+ * Writes a buffer or string to stdout, only if stdin is a tty.
+ * @param {NodeJS.ArrayBufferView} b The outpus which should be written to stdout (if stdin is a tty).
+ */
+function writeTTYOutput(b) {
+  if (isTTY) {
+    writeOutput(b);
+  }
+}
+
+/**
+ * If stdin is a tty, ensure raw mode is on prior to executing cb,
+ * then reset it to its previous state afterwards.  If stdin isn't a tty,
+ * cb will just be executed normally.
+ *
+ * @template T
+ * @param {() => T} cb The function which is being executed.
+ * @return {T} The return value of cb
+ */
+function ensureRawMode(cb) {
+  if (!isTTY) {
+    return cb();
+  }
+
+  const previousRawMode = process.stdin.isRaw;
+
+  try {
+    process.stdin.setRawMode(true);
+
+    return cb();
+  } finally {
+    process.stdin.setRawMode(previousRawMode);
+  }
```

Comment:
Finally always messes with my head when it comes to flow control, I find it hard to reason about.

Wonder if it's easier to read if this function returned a callback to reset to previous raw mode.

```js
function question(prompt) {
  const resetRawMode = ensureRawMode();
  ...promptcode...
  resetRawMode();
  return str;
}
```

Reply:
I added the finally + made it like this for 2 reasons:
1. We should reset it even if something throws (but throw the error anyways)
2. Its easy to miss restoring the tty's state.  This is the closest JS equivalent to golang's defer that I can think of.

Reply:
I'll move the return outside of the try/finally block, since I can see how that gets confusing.

Reply:
Yeah that's better, thanks